docs/getting-started/using-gnokey.md

```diff
@@ -0,0 +1,777 @@
+
+```go
+package main
+
+import "gno.land/r/demo/userbook"
+
+func main() {
+  println(userbook.SignUp())
+}
+```
+
+Now we will be able to provide this to the `maketx run` subcommand:
+```bash
+gnokey maketx run \
+-gas-fee 1000000ugnot \
+-gas-wanted 20000000 \
+-broadcast \
+-chainid portal-loop \
+-remote "https://rpc.gno.land:443" \
+mykey ./script.gno
+```
+
+After running this command, the chain will execute the script and apply any state
+changes. Additionally, by using `println`, which is only available in the `Run`
+& testing context, we will be able to see the return value of the function called.
+
+#### The power of `Run`
+
+Specifically, the above example could have been replaced with a simple `maketx call`
+call. The full potential of run comes out in three specific cases:
+1. Calling realm functions multiple times in a loop
+2. Calling functions with non-primitive input arguments
+3. Calling functions with receiver objects
+
+Let's look at each of these cases in detail. To demonstrate, lets use the
+following example realm which we will call:
+
+```go
+package foo
+
+import "gno.land/p/demo/ufmt"
+
+var (
+	MainFoo *Foo
+	foos    []*Foo
+)
+
+type Foo struct {
+	bar string
+	baz int
+}
+
+func init() {
+	MainFoo = &Foo{bar: "mainBar", baz: 0}
+}
+
+func (f *Foo) String() string {
+	return ufmt.Sprintf("Foo - (bar: %s) - (baz: %d)\n\n", f.bar, f.baz)
+}
+
+func NewFoo(bar string, baz int) *Foo {
+	return &Foo{bar: bar, baz: baz}
+}
+
+func AddFoos(multipleFoos []*Foo) {
+	foos = append(foos, multipleFoos...)
+}
+
+func Render(_ string) string {
+	output := ""
+
+	for _, f := range foos {
+		output += f.String()
+	}
+
+	return output
+}
+```
+
+This realm is deployed to [`gno.land/r/leon/run/examples/foo`](https://gno.land/r/leon/run/examples/foo)
+on the Portal Loop testnet.
+
+1. Calling realm functions multiple times in a loop:
+```go
+package main
+
+import (
+  "gno.land/r/leon/run/examples/foo"
+)
+
+func main() {
+  for i := 0; i < 5; i++ {
+    println(foo.Render(""))
+  }
+}
+```
+
+2. Calling functions with non-primitive input arguments:
+
+Currently, `Call` only supports primitives for arguments. With `Run` these
+limitations are removed - we can execute a function that takes in a struct, array,
+or even an array of structs.
+
+We are unable to call `AddFoos()` with the `Call` message type, while with `Run`,
+we can:
+
+```go
+package main
+
+import (
+  "gno.land/r/leon/run/examples/foo"
+  "strconv"
+)
+
+func main() {
+  var multipleFoos []*foo.Foo
+
+  for i := 0; i < 5; i++ {
+    newFoo := foo.NewFoo(
+      "bar"+strconv.Itoa(i),
+      i,
+    )
+
+    multipleFoos = append(multipleFoos, newFoo)
+  }
+
+  foo.AddFoos(multipleFoos)
+}
+
+```
+
+3. Calling functions with receiver objects:
+
+```go
+package main
+
+import "gno.land/r/leon/run/examples/foo"
+
+func main() {
+	println(foo.MainFoo.String())
+}
+```
+
+Finally, we can call functions that are on top-level objects, which is not possible
```

Comment:
Is this different than calling methods on realm objects? It seems the same to me

Reply:
No, it's not the same. Hopefully the example makes it clear, but the idea is that you can have a top-level exported object in a realm, which has a receiver method on it. You can call that method via `Run`, like so:

RealmA:

```go
package realmA

type MyType struct {...}

var Exported MyType = {...}

func (obj *MyType) MyFunc() {...}

func init() {...}

```
And somewhere else, for `MsgRun`:
```go
package main

import "realmA"

func main() {
 realmA.Exported.MyFun()
}

```

Reply:
I think we are saying the same thing. "functions with receiver objects" just sounds weird and is not really something people say. Something like "methods on exported realm objects" would be better in my opinion.

Reply:
I was thinking on this more. The phrasing as it exists now "with receiver objects" is incorrect. You can call a method on a value with a non-pointer receiver object and still change the state.  For example:
```go
package bar

type S struct {
	t *T
}

type T struct {
	counter int
}

var Obj S = S{t: new(T)}

func (s S) IncCounter() {
	s.t.counter++
}

func GetCounter() int {
	return Obj.t.counter
}
```
In this case, calling `bar.Obj.IncCounter()` will result in T's counter state being changed even though the method's receiver is not a pointer.

Reply:
Got it from the call today; it was the terminology. [methods](https://go.dev/tour/methods/1) vs functions. switched it out: [aa26e6a](https://github.com/gnolang/gno/pull/2226/commits/aa26e6a9db9c6b0e53141c8178f156ec3f41058f)